src/Type/Php/JsonThrowOnErrorDynamicReturnTypeExtension.php

```diff
@@ -99,4 +119,73 @@ private function isBitwiseOrWithJsonThrowOnError(Expr $expr, Scope $scope): bool
 			$this->isBitwiseOrWithJsonThrowOnError($expr->right, $scope);
 	}
 
+	private function narrowTypeForJsonDecode(FuncCall $funcCall, Scope $scope): Type
+	{
+		$args = $funcCall->getArgs();
+		$isForceArray = $this->isForceArray($funcCall);
+
+		$firstArgValue = $args[0]->value;
+		$firstValueType = $scope->getType($firstArgValue);
+
+		if ($firstValueType instanceof ConstantStringType) {
+			$resolvedType = $this->resolveConstantStringType($firstValueType, $isForceArray);
+		} else {
+			$resolvedType = new MixedType();
+		}
+
+		// prefer specific type
+		if (! $resolvedType instanceof MixedType) {
+			return $resolvedType;
+		}
+
+		// fallback type
+		if ($isForceArray) {
+			return new UnionType([
```

Comment:
As I already said in https://github.com/phpstan/phpstan-nette/pull/89#discussion_r795243535, we really can't use a union. Right now people using `json_decode` are "penalized" on level 9 thanks to `mixed`. With this precise union, they'd be penalized on level 7. That's not something we can do in a minor version, maybe ever.

The best we can do here I think is that we can return `MixedType` for case with `$assoc=false` and `MixedType minus stdClass` for `$assoc=true`.

Reply:
I've update the fix :+1: